src/nni_manager/training_service/pai_base/pai/paiTrainingService.ts

```diff
@@ -0,0 +1,326 @@
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+ * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+'use strict';
+
+import * as cpp from 'child-process-promise';
+import * as fs from 'fs';
+import * as path from 'path';
+// tslint:disable-next-line:no-implicit-dependencies
+import * as request from 'request';
+import * as component from '../../../common/component';
+
+import { Deferred } from 'ts-deferred';
+import { String } from 'typescript-string-operations';
+import {
+    HyperParameters, NNIManagerIpConfig, TrainingService,
+    TrialJobApplicationForm, TrialJobDetail, TrialJobMetric
+} from '../../../common/trainingService';
+import { delay, generateParamFileName,
+    getExperimentRootDir, getIPV4Address, getVersion, uniqueString, unixPathJoin } from '../../../common/utils';
+import { CONTAINER_INSTALL_NNI_SHELL_FORMAT } from '../../common/containerJobData';
+import { TrialConfigMetadataKey } from '../../common/trialConfigMetadataKey';
+import { execMkdir, validateCodeDir, execCopydir } from '../../common/util';
+import { PAI_LITE_TRIAL_COMMAND_FORMAT } from './paiData';
+import { NNIPAITrialConfig } from './paiConfig';
+import { PAIJobRestServer } from './paiJobRestServer';
+import { PAIBaseTrainingService } from '../paiBaseTrainingService';
+import { PAIBaseClusterConfig, PAIBaseTrialJobDetail } from '../../pai_base/paiBaseConfig';
+
+const yaml = require('js-yaml');
+
+/**
+ * Training Service implementation for OpenPAI (Open Platform for AI)
+ * Refer https://github.com/Microsoft/pai for more info about OpenPAI
+ */
+@component.Singleton
+class PAITrainingService extends PAIBaseTrainingService {
+    protected paiTrialConfig: NNIPAITrialConfig | undefined;
+
+    constructor() {
+        super();
+    }
+
+    public async run(): Promise<void> {
+        this.log.info('Run PAI training service.');
+        const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
+        await restServer.start();
+        restServer.setEnableVersionCheck = this.versionCheck;
+        this.log.info(`PAI Training service rest server listening on: ${restServer.endPoint}`);
+        await Promise.all([
```

Comment:
This `run` is almost same as `PAIYarnTrainingService.run`,  should manage to move it up to base class

Reply:
fixed.